### PR TITLE
Reader Improvements Phase 3: Adds recommended sites in the Discover feed

### LINKED ISSUE #14766: Reader Improvements: Sites/Blogs you might like
**Project Issue:** https://github.com/wordpress-mobile/WordPress-iOS/issues/14768

Screen Shot 2020-08-26 at 2 35 18 PM

### Goal
Display a list of recommended sites to follow in the Discover feed, and allow the user to follow a site directly in the feed or tap it to see all the posts for the site. 

### Designs
- [Explorations](https://www.figma.com/file/NOw2XrKMsWc5RwBF6Flax2/Sites-you-might-like?node-id=2%3A127)
- [Specs](https://www.figma.com/file/RjrZZjXCtNYXWsPt5i0jha/Sites-You-Might-Like?node-id=16%3A435)

### PR DESCRIPTION
**Pull request by emilylaguna**
Part of project: #14768
Fixes #14766

#### Light Mode

| Not Following | Following |
|:---:|:---:|
|||

#### Dark Mode
| Not Following | Following |
|:---:|:---:|
|||

### To test:
1. Launch the app

#### Expectation: The app launches and you are not asked to sign in again (CoreData migration check)

2. Tap on Reader
3. Tap on Discover
4. Scroll down until you see the Sites to follow section
5. Tap on the following icon

#### Expectation: The icon turns gray, and you see a "you followed this site" toast message

6. Tap the following icon again

#### Expectation: The icon turns pink

7. Tap on the site

### PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by emilylaguna** [comment]
This may also fix the issue of cards being cut off: https://github.com/wordpress-mobile/WordPress-iOS/issues/14885

Specifically this commit: https://github.com/wordpress-mobile/WordPress-iOS/pull/15116/commits/ab81161156803cb89c33a389b7c4bbcbd8c4c8e3

File: WordPress/Classes/Models/ReaderCard+CoreDataClass.swift [hunk 1]
```diff
@@ -13,17 +14,25 @@ public class ReaderCard: NSManagedObject {
             return .post
         }
 
-        if topics != nil {
+        if topicsArray.count != 0 {
```
Comment (yaelirub) [inline WordPress/Classes/Models/ReaderCard+CoreDataClass.swift]: Shouldn't this be `topicsArray.count > 0`?

File: WordPress/Classes/Models/ReaderCard+CoreDataClass.swift [hunk 2]
```diff
@@ -13,17 +14,25 @@ public class ReaderCard: NSManagedObject {
...
             return .topics
         }
 
+        if sitesArray.count != 0 {
```
Comment (yaelirub) [inline WordPress/Classes/Models/ReaderCard+CoreDataClass.swift]: Shouldn't this be sitesArray.count > 0?

**Comment by leandroalonso** [comment]
@emilylaguna the separator between both the sites you might like and topics you might like is gone in light mode (iPhone 11 14.0):

1) Separator gone

| Not Following | Following |
|:---:|:---:|
|Simulator Screen Shot - iPhone 11 - 2020-10-16 at 09 40 49|Simulator Screen Shot - iPhone 11 - 2020-10-16 at 09 40 57|

When I switch to dark mode, however, it appears correctly.

2) Tap delay

When I tap one of the sites, there is a delay between my tap and the screen finally opening. Have you experienced something similar?

3) Margin between the title and the blogs

I guess we could add the same margin between the title and the content, same used in "Topics you might like"

4) Events

We want to know how is the "sites you might like" performing, both in terms of users visiting the site and following them. I guess we can add events for those two things, wdyt?

**Comment by mbshakti** [comment]
Already found some sites to follow :) Some comments:

+1 to @leandroalonso's comment, can we had some margin between "sites to follow", and the sites list? I think we are using either 12px or 16px for the "Topics you might like", lets use the same here. 

I'm also experiencing some delay after tapping before being redirected to the site. 

Bug I encountered on iPhone & iPad:
If I tap on any recommended site, go to the site page, and tap the follow button from the site header, my app crashes. This happened each time on ipad & phone, took a recording below from iPad. (version 13.5.1 on iPad and 14 on iPhone)

Kapture 2020-10-16 at 10 27 16

**Comment by emilylaguna** [comment]
> When I tap one of the sites, there is a delay between my tap and the screen finally opening. Have you experienced something similar?

@leandroalonso @mbshakti 
I don't see any delay when tapping on the site on my end. I have been getting random delays elsewhere but I think it's due to the hundreds of logs we're getting from iOS 14 and CoreData.

I can confirm using breakpoints that there is no delay from the time I tap the button to the time the push is triggered.

**Comment by emilylaguna** [comment]
@leandroalonso Another look please 🥺

**Review by leandroalonso — APPROVED** [review]
> I don't see any delay when tapping on the site on my end.

The issue I'm having is probably simulator related, even rotating the device is being very slow.

Besides that, everything looks good to me. :shipit: 

@yaelirub do you want to double-check before we merge?

**Comment by emilylaguna** [comment]
Added the iPad follow button

Simulator Screen Shot - iPad Air (3rd generation) - 2020-10-16 at 14 31 22

**Review by leandroalonso — APPROVED** [review]

File: WordPress/Classes/Models/ReaderSiteInfoSubscriptions.swift [hunk 3]
```diff
@@ -5,6 +5,22 @@ import CoreData
...
     @NSManaged open var siteTopic: ReaderSiteTopic
     @NSManaged open var sendPosts: Bool
+
+    class func createOrUpdate(from remoteSiteInfo: RemoteReaderSiteInfo, topic: ReaderSiteTopic, context: NSManagedObjectContext) -> ReaderSiteInfoSubscriptionPost? {
+        guard remoteSiteInfo.postSubscription.wp_isValidObject() else {
+            return nil
+        }
+
+        var subscription = topic.postSubscription
+        if !(subscription?.wp_isValidObject() ?? false) {
```
Comment (yaelirub) [inline WordPress/Classes/Models/ReaderSiteInfoSubscriptions.swift]:
```suggestion
        if subscription?.wp_isValidObject() == false {
```

File: WordPress/Classes/Models/ReaderSiteInfoSubscriptions.swift [hunk 4]
```diff
@@ -13,4 +29,22 @@ import CoreData
...
     @NSManaged open var sendComments: Bool
     @NSManaged open var postDeliveryFrequency: String
+
+    class func createOrUpdate(from remoteSiteInfo: RemoteReaderSiteInfo, topic: ReaderSiteTopic, context: NSManagedObjectContext) -> ReaderSiteInfoSubscriptionEmail? {
+        guard remoteSiteInfo.emailSubscription?.wp_isValidObject() ?? false else {
+            return nil
+        }
+
+        var subscription = topic.emailSubscription
+        if !(subscription?.wp_isValidObject() ?? false) {
```
Comment (yaelirub) [inline WordPress/Classes/Models/ReaderSiteInfoSubscriptions.swift]:
```suggestion
        if subscription?.wp_isValidObject() == false {
```

File: WordPress/Classes/ViewRelated/Reader/ReaderTopicsCardCell.swift [hunk 5]
```diff
@@ -1,166 +1,37 @@
...
-        return self.contentSize
+        return cell
     }
 
-    override var contentSize: CGSize {
-        didSet {
-            self.invalidateIntrinsicContentSize()
-        }
+    private enum Constants {
+        static let title = NSLocalizedString("You might like", comment: "A suggestion of topics the user might ")
```
Comment (yaelirub) [inline WordPress/Classes/ViewRelated/Reader/ReaderTopicsCardCell.swift]: Might...? missing like in the comment

**Review by yaelirub — APPROVED** [review]
Brilliant! 👏 👏 👏

**Pull request merged**